test/Node/PackageLockSpec.hs

```diff
@@ -329,6 +373,26 @@ buildGraphSpec testDir =
         ]
         graph
 
+    -- What we're really trying to test here is that buildGraph finishes its
+    -- work in a reasonable amount of time. If buildGraph starts to recurse
+    -- infinitely, eventually that thread will be forced to stop and we'll
+    -- signal a failure. This style of test is prone to be flakey, so
+    -- it may need to be removed/turned off in the future.
+    it "Should avoid recursing into peerDep cycles" $
```

Comment:
I'm presuming that without this PR change - this case would be stuck in an infinite loop, or _just_ extremely poor performance?

If it's the former then I'm not in favor of adding threading/timing into this test case. I know this is the pragmatic approach, but I would prefer if we have more of a pure unit test. Since your test case is not changing (i.e. `peerDepCycle`), I rather that we just assert the graph's correctness.

Reply:
sigh! Letting the unit test run infinitely is an a-ok signal of failure (I'm personally a-ok with this) but, it is not ideal (hoard up CI). Another approach would be to introduce some form of a `MaybeCircuitBreaker` into the graphing effect - but this seems out of scope for this PR. Can you add this as a tech debt (e.g. Identify a better way to test/break infinite loops in graphing effect) into our backlog?

Reply:
ok, I will definitely add that and reference this conversation. Thanks!

Reply:
The long term solution is to form the graph early, in a way that does not infinitely recurse (the `dependencies` field is recursive, but only because the data is physically nested, so the recursion there is not infinite).  Then, if needed, we can check for cycles in the graph using existing `AdjacencyMap` utils.

Reply:
Data-traversal recursion without jumping is basically the only long term solution here.  This prevents us from even having to test for cycles.

Reply:
I'm wondering now if it was a mistake for `addPeerDeps` to recurse at all.

If we can assume that all peer dep versions are also represented in the `dependencies` object, couldn't `addPeerDeps` just make edges and the rest of the information to be added when the top-level traversal gets to it? I'm already making the above assumption in `addPeerDeps` because `addPeerDeps` looks up the corresponding dependency object in order to recurse into it.

If the above works, we don't need to look up whether a vertex has been added and the peer deps support becomes less invasive.

Reply:
If you can think of a solution that doesn't use jumping recursion for peer deps, that's worth doing now.